### PR TITLE
Don't get client state until need it

### PR DESCRIPTION
**Pull request by martindurant**
Seems to fix test.
I'm not sure if you can specify a type for cython and set/compare to `None` like this. That pattern seems to happen in a number of places, including the kwargs of `update_graph`.

I note that on L3991, variable `ts` is created, but not used. This isn't being picked up by flake.

### CODE REVIEW

File: distributed/scheduler.py [hunk 1]
```diff
@@ -3933,7 +3935,7 @@ def update_graph(
             if tasks[k] is k:
                 del tasks[k]
 
-        cs: ClientState = parent._clients[client]
+        cs: ClientState = None  # parent._clients[client]
```
Comment (jakirkham) [inline distributed/scheduler.py:3938]: We also make use of `cs` below

https://github.com/jakirkham/distributed/pull/3/files#diff-bbcf2e505bf2f9dd0dc25de4582115ee4ed4a6e80997affc7b22122912cc6591R4190

Comment (martindurant) [reply distributed/scheduler.py:3938]: Only if a task ended up in `touched_tasks`, which means we must have gone past the place where we set its value, no?

**Comment by jakirkham** [comment]
Thanks Martin 🙂

Yeah `None` is allowed by basically all builtin and Cython extension types. I think they may change that in the future in Cython 3, but don't know when that will come out.

There is certainly quite a bit happening in this function. Like `cs` gets used again many lines later. `ts` is used in a few spots (though agree that particular line is odd)

**Comment by martindurant** [comment]
> None is allowed by basically all builtin and Cython extension types

So in a real .pyx file, we would certainly save time by using direct API calls like https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItemString ? I imagine this isn't possible in pure-python mode.

**Comment by jakirkham** [comment]
It means that Cython will inject some checks to see if variables are `None`. Since `None` is defined at interpreter launch and so has a fixed pointer, this is a fast pointer comparison (analogous to checking for `NULL` in C code). It will then following it with a call to `PyDict_SetItem`. So the only difference in time is this `None` check, which is pretty cheap

Agree it would be nice to skip the `None` check altogether. IIUC this is what Cython 3 does (though that's still in alpha)

**Comment by martindurant** [comment]
OK, the None may be a red-herring, at least my suggestion in this PR is legit. Do you think in the end it's useful? Are there indeed cases where the `cs` is needed further down the function, but not set at the point I wrote?

**Pull request closed without merging**